### ISSUE TITLE
cardano-ledger does not build with current version of streaming-bytestring

### ISSUE THREAD

**Issue by mark-stopka** [IntersectMBO/cardano-ledger#1973]
[cardano-ledger](https://github.com/input-output-hk/cardano-ledger-specs/tree/master/byron/ledger/impl) does not build with 0.2.0 of [streaming-bytestring](https://hackage.haskell.org/package/streaming-bytestring), due to change in API, building with `development` Cabal flag

```
flag development
    description: Disable `-Werror`
    default: False
    manual: True
```

is certainly not a correct way to solve the issue, we have a quick & dirty patch available here, feel free to use it as is or refactor further as required...

https://build.opensuse.org/package/view_file/devel:languages:haskell:cardano/ghc-cardano-ledger/fix-deprecations.patch?expand=1

FYI & credits: @mimi1vx

**Comment by mrBliss** [comment]
According to the "Uploaded" field on Hackage, `streaming-bytestring-0.2.0` has been released on 2020-10-26. Since we use an older Hackage `index-state` for [`cardano-ledger-specs`](https://github.com/input-output-hk/cardano-ledger-specs/blob/29527dec11fe1689ffe5cbb3960ff2f2327d4b6f/cabal.project#L1) (and [`cardano-node`](https://github.com/input-output-hk/cardano-node/blob/62485960494d914f8efd06ed0d8357d41a8f9d26/cabal.project#L1)), Cabal doesn't pick up the newer version and we don't see this error.

Either we switch to `0.2.0`, requiring code changes and `index-state` updates in all downstream repos, or we add a `< 0.2.0` bound to the [`streaming-bytestring` dependency](https://github.com/input-output-hk/cardano-ledger-specs/blob/d5eaac6c4b21a8e69dc3a5503a72e3c3bfde648e/byron/ledger/impl/cardano-ledger.cabal#L156).

**Comment by mark-stopka** [comment]
> we add a < 0.2.0 bound to the streaming-bytestring dependency.

This just complicates things even more for us as a package maintainers, because we will have to add additional macro to override your constrain you place in the `cabal` file...

**Comment by mrBliss** [comment]
> > we add a < 0.2.0 bound to the streaming-bytestring dependency.
> 
> This just complicates things even more for us as a package maintainers, because we will have to add additional macro to override your constrain you place in the `cabal` file...

I don't understand, why would you override the `< 0.2.0` constraint if `cardano-ledger-specs` doesn't support `0.2.0`, it won't build.

**Comment by mark-stopka** [comment]
> > > we add a < 0.2.0 bound to the streaming-bytestring dependency.
> > 
> > 
> > This just complicates things even more for us as a package maintainers, because we will have to add additional macro to override your constrain you place in the `cabal` file...
> 
> I don't understand, why would you override the `< 0.2.0` constraint if `cardano-ledger-specs` doesn't support `0.2.0`, it won't build.

Because we have patched it here as I said in my initial issue post? As a distribution package maintainers we can't just ignore upstream API changes willy-nilly, we may downgrade `streaming-bytestring` distribution wide, which would on the other hand break other software that already migrated to the new API.

https://build.opensuse.org/package/view_file/devel:languages:haskell:cardano/ghc-cardano-ledger/fix-deprecations.patch?expand=1

**Comment by mark-stopka** [comment]
> > > > we add a < 0.2.0 bound to the streaming-bytestring dependency.
> > > 
> > > 
> > > This just complicates things even more for us as a package maintainers, because we will have to add additional macro to override your constrain you place in the `cabal` file...
> > 
> > 
> > I don't understand, why would you override the `< 0.2.0` constraint if `cardano-ledger-specs` doesn't support `0.2.0`, it won't build.
> 
> Because we have patched it here as I said in my initial issue post? As a distribution package maintainers we can't just ignore upstream API changes willy-nilly, we may downgrade `streaming-bytestring` distribution wide, which would on the other hand break other software that already migrated to the new API.
> 
> https://build.opensuse.org/package/view_file/devel:languages:haskell:cardano/ghc-cardano-ledger/fix-deprecations.patch?expand=1

To further add on that, if you believe our patch is insufficient, and should be improved, and that there are other areas in the overall `cardano-node` codebase that may require change, we are more than willing to put in the work if you don't have the bandwidth to do it, however we would appreciate the assistance from IOG team, as we are fairly new to the codebase and have just high-level understanding which we of course continue to improve over time, as we work on issues such as this one.

**Comment by dcoutts** [comment]
Thanks for the notice. We'll add the constraint so that our .cabal files reflect reality. At some point we'll update to newer libs, but there is no rush on this.

> As a distribution package maintainers we can't just ignore upstream API changes willy-nilly, we may downgrade streaming-bytestring distribution wide, which would on the other hand break other software that already migrated to the new API.

I also used to be a maintainer of Haskell packages in an OS distribution. It is not in general possible to get all packages to agree on a single common set of versions of all Haskell libraries, and trying to force it is like trying to boil the ocean. It is a long-settled issue in the Haskell packaging world that in general it is necessary to support multiple versions of libraries.

Yes aiming towards consistency is good, but trying to enforce it at all times is simply not possible.

**Comment by mark-stopka** [comment]
> At some point we'll update to newer libs, but there is no rush on this.

Are you just going to ignore the aforementioned:

> To further add on that, if you believe our patch is insufficient, and should be improved, and that there are other areas in the overall cardano-node codebase that may require change, we are more than willing to put in the work if you don't have the bandwidth to do it, however we would appreciate the assistance from IOG team, as we are fairly new to the codebase and have just high-level understanding which we of course continue to improve over time, as we work on issues such as this one.

We have an expert estimate by a leading 3rd party software maintanance company, that it would take 12 months of due-diligence to overtake the development of `cardano-node` as a whole, we have 3 months of secured funding, we would like to narrow that gap in cooperation with IOG and it's sub-contractors such as Well-Typed and we are offering a free labour, yet the only reaction for our offer we get is "we are in no rush in reduction of technical debt"?

**Comment by mark-stopka** [comment]
This actually fixed the root-cause https://github.com/input-output-hk/cardano-ledger-specs/pull/2020, thx @newhoggy!

**Comment by dcoutts** [comment]
@mark-stopka we're very happy to accept PRs for small issues like this.

The "we're in no rush" refers to us spending time on these things, when there are higher priority issues. If you want to contribute your time, we're grateful to accept the contributions.

And specifically "no rush" when it comes to the package version upgrade treadmill. This is a never ending thing, and the answer is not to spend all of one's time trying to be at the bleeding edge all the time. It is far more efficient to update from time to time in bigger jumps.

**Comment by mark-stopka** [comment]
@dcoutts, thank you for your reply, I surely undarstand that point of view also, this effort we are undergoing right now is part of a bigger picture I cannot speak to at this point in time, however I will bring it up with the project steering group on our next meeting and we will try to find a viable solution.

For us as a project, we will probably clasify dependencies into two categories "Cardano Only" and "General Haskell", and as maintainers, maintenance and support team we will _have to_ push for quick upgrades (and will invest the required time & material) in libraries that are in the **General Haskell** library and be able to provide a lot of flexibility with the _Cardano Only_ dependencies (even if they are from Hackage but are not used by other critical Haskell applications within the distribution).

We are currently developing the test-suite, and automating the release management process for Cardano bits and pieces, I should note here, that we have 2 versions, one is rolling updates and without support (there we can live with "it does not build for now"), but every now and then the rolling version is frozen, broken builds get fixed during 3 - 6 months freeze - hence that would be our window of opportunity to update any **General Haskell** dependencies and new release of openSUSE Leap + SUSE Linux Enterprise is cut.

With that said, we have noticed quite a few pieces that we think would be worth opening up to the whole world on Hackage which are currently living deeply in the bowels (I don't remember what TV show this joke is from, I am sorry, don't take it offensively ;) ) of IOHK repositories such as `goblins` for instance, so it may be worth opening a dialogue about that also. This would bring added value even for IOG as the packages that make it into SLE undergo random independent security audits, follown strict vulnarability management processes and have comprehensive integration testing, especially when it comes to those planned to be deployed on System Z which is the case of `cardano-node`.

TL;DR been there, done that (with many projects - incl. Python and you know what a mess `pip` is, or PERL CPAN,...), we've always managed to find a solution that works for everyone, and I strongly believe it will be the same case here, as the time progresses and we will be able to share more details.

In addition, sorry for little aggresive tone (in my circles it would not be considered such as the governance & compliance field is the way it is), but you may notice end of month is close, and beginning of a month is when that project steering comittee meeting takes place, and I don't like to tell excuses, but rather report progress... :) which now I can, so thanks a milion!

For now, you can consider us a watchers of future API breaking dependency upgrades who may also at times come along with patches that fix such breaking API changes - I now assume, you would prefer a patch that also updates `Hackage` index state due to the way you do your CI and releases, so we will keep that in mind for the future, does that sound suitable?
